### ISSUE TITLE
Python Type Hints

### ISSUE THREAD

**Issue by gitpushdashf** [grpc/grpc#29041]
It looks like Python has type hints for aio, but not all of the code. `py.typed` also isn't shipped, which would mark the wheel as having type hints. This results in external type hints, like this: https://github.com/shabbyrobe/grpc-stubs

It'd be great if we could have type hints for the Python grpc library.

**Comment by lidizheng** [comment]
We might need volunteers to own this item. The type hints used to be blocked by 27 deprecation, and Python 35 (lacking variable annotations). Infrastructure-wise, both blockers are cleaned now. This will be a significant improvement for gRPC Python's development workflow.

**Comment by madkinsz** [comment]
I'm new to this project, but I'm interested in taking a look at this. It seems like it'd be easy to follow the style for the type hints already introduced in the `aio` implementation, but maybe it'd be good to get some guidelines established. For example:
- Type hints should be specified inline vs a separate stub file.
- Generic type variables and common definitions should be stored in a `_typing` module e.g. [`aio._typing`](https://github.com/grpc/grpc/blob/master/src/python/grpcio/grpc/aio/_typing.py).
- Optional types should be explicit e.g. `Optional[int] = None` instead of `int = None`
- Bleeding edge Python type features should not be used e.g. [| for unions](https://peps.python.org/pep-0604/)

Are these types currently checked for accuracy by any automations?

It seems ideal to introduce type hints gradually in small reviewable pieces, perhaps by module? I'm curious if there are suggestions. Which modules are the most important for users?

I noticed the AsyncIO gRPC design includes [a section on generated code](https://github.com/lidizheng/proposal/blob/grpc-python-async-api/L58-python-async-api.md#introduce-typing-to-generated-code), is that a separate goal?

**Comment by yingshaoxo** [comment]
type hint is something that needed

___

<img width="1196" alt="image" src="https://user-images.githubusercontent.com/17190829/209931180-7fadf2d5-67ad-4088-90c0-86ea6fcbc00e.png">
<img width="1014" alt="image" src="https://user-images.githubusercontent.com/17190829/209931213-61508c48-9976-4d1a-a705-163d25088040.png">

___

Why for python, there got so many errors with grpc?

**Comment by gnossen** [comment]
Type hints for the sync stack (i.e. non-asyncio stubs) are being actively worked on and should be completed within the next month or so. @XuanWang-Amos to provide more detail on that.

@yingshaoxo Based on your screenshot, your issue is in your generated message code, not stub code. You'll just need to add the `--pyi_out` argument to your invocation of `grpc_tools.protoc` to generate type stubs.

**Comment by XuanWang-Amos** [comment]
Yes, I'm currently working on this and we already added types for most of the internal files, we'll also publish a gRFC soon (will link here too).

**Comment by cprice404** [comment]
can't wait for this!  Thank you!

**Comment by dgcnz** [comment]
Is there any update?

**Comment by XuanWang-Amos** [comment]
I'm still working on this, but it was deprioritized this quarter due to some other urgent work. I expect to complete it in the next quarter.

**Comment by Dreamsorcerer** [comment]
As long as the annotations are not _wrong_, it would be useful to start exposing them as soon as possible, even if they are only partially complete. Might encourage more people to contribute some missing annotations.

**Comment by XuanWang-Amos** [comment]
Unfortunately we're not able to finish this in this quarter, it's now delayed to next quarter.

We do have a PR in review for anyone who wants to know the progress: https://github.com/grpc/grpc/pull/31769

**Comment by shohei-ihaya** [comment]
Any updates on this?

**Comment by XuanWang-Amos** [comment]
Sorry again about the delay. Unfortunately, this work has been deprioritized again in favor of gRPC Python observability work. We will update this issue once we begin working on it.

**Comment by Dreamsorcerer** [comment]
I would like to refer to my previous comment: https://github.com/grpc/grpc/issues/29041#issuecomment-1584898751
There seems to be no reason to delay exposing the annotations for years while it gets deprioritised, just because they are not 100% complete. We have dozens of typing hacks and ignores in place (mostly caused by #31880) that would be resolved by simply adding an empty `py.typed` file to the library.

**Comment by XuanWang-Amos** [comment]
Hi, thank you for your suggestion.

From my understanding, adding a `py.typed` file indicates that a package supports type checking. However, in our case, this isn't accurate. Therefore, I have some concerns regarding the potential implications of adding this file without proper types.

I do agree this task has been delayed for a long time now, I'll bring this up in our team meeting and try get it prioritized.

**Comment by Dreamsorcerer** [comment]
> From my understanding, adding a `py.typed` file indicates that a package supports type checking.

Yes, it indicates that it has type annotations available for type checking, and so mypy etc. will start using those annotations for type checking.

> However, in our case, this isn't accurate.

It doesn't mean you have 100% perfect annotations everywhere. It just means that there are enough annotations to provide something useful (which with #31880, it certainly will be even if nothing else is typed). If some modules are missing type annotations, that is functionally the same as if they had `Any` added everywhere, so mypy will just treat them as dynamically typed and not check them. The missing ones can then be added gradually in later releases.

As an example, sqlalchemy have done this very well, they focused on adding annotations to the most common user-facing parts of the library and have been gradually filling in the rest of the annotations as they go. Despite exposing annotations to users years ago, they still appear to have ~20 modules untyped (https://github.com/sqlalchemy/sqlalchemy/issues/6810).

**Comment by XuanWang-Amos** [comment]
Thank you for clarifying that. We'll discuss this internally and create a PR if the team agrees.

**Comment by XuanWang-Amos** [comment]
Hi, I don't think it's safe to add `py.typed` at this point for two reasons:
  1. While mypy skips analyze functions if it's missing type hints in the signature, it will start checking for classes. For example, we don't have annotation for [HandlerCallDetails](https://github.com/grpc/grpc/blob/master/src/python/grpcio/grpc/__init__.py#L1361), mypy will start raising `attr-defined` error if [user tries to access HandlerCallDetails.method in their code](https://github.com/grpc/grpc/blob/master/examples/python/auth/async_customized_auth_server.py#L53) even when we know `HandlerCallDetails` have `method` attribute.
  2. We have some errors (Like the one mentioned in https://github.com/grpc/grpc/issues/31880) in existing type hints, it will start popping up once type check is enabled.

To better prioritize the type hint work in the next quarter, please comment on or react to this issue. We would greatly appreciate your input.

**Comment by Dreamsorcerer** [comment]
>     1. While mypy skips analyze functions if it's missing type hints in the signature, it will start checking for classes. For example, we don't have annotation for [HandlerCallDetails](https://github.com/grpc/grpc/blob/master/src/python/grpcio/grpc/__init__.py#L1361), mypy will start raising `attr-defined` error if [user tries to access HandlerCallDetails.method in their code](https://github.com/grpc/grpc/blob/master/examples/python/auth/async_customized_auth_server.py#L53) even when we know `HandlerCallDetails` have `method` attribute.

Right, that could be a bit annoying. But, could also be focused on first, possibly with a quick hack which just defines the classes as `Any` when type checking.

e.g.
```
if TYPE_CHECKING:
    MyClass = Any
else:
    class MyClase:
        ...
```

>     2. We have some errors (Like the one mentioned in [Wrong return type for abort() #31880](https://github.com/grpc/grpc/issues/31880)) in existing type hints, it will start popping up once type check is enabled.

Uhh, no, the opposite. This is why I keep asking. We have type errors _now_ from #31880, _because_ the type annotations are not exposed. Exposing the type annotations _fixes_ these type errors today.

**Comment by cprice404** [comment]
Just replying again in response to @XuanWang-Amos 's solicitation - anything that can be done to move the ball forward on some basic mypy checking would be really valuable and appreciated. Working with the async APIs vs the sync APIs at the moment is a night and day difference.

**Comment by chrisprobst** [comment]
Following this thread forever and highly interested in getting this landed! +1 +1 +1

**Comment by jeffpicard** [comment]
This would be great for our company as well!

**Comment by azryve** [comment]
I have tried to use grpc.aio for the first time this week and it has been quite a challenge, and I had a relatively easy case with small codebase, essentially a poc. I dread to think about something more complicated, evolving over time tbh.

**Comment by iamWing** [comment]
Following this thread +1

Started to use grpc with Python recently but so far it's painful trying to figure out the types and annotate the type info in my code base using `grpc` & `grpc_testing` packages.

I have my code base fully type hinted and it works great in vscode with pylance and copilot until the moment I'm using a package that doesn't come with type hints. It'd be so helpful if type hint is coming soon.

**Comment by brianoflondon** [comment]
Also adding a vote. This is pretty essential in this day and age.

**Comment by sreenithi** [comment]
We are currently working on adding typings to the gRPC Python Repo.

**Comment by andrew-cho-architect** [comment]
+1

**Comment by creyD** [comment]
@sreenithi Yes please update on the state, and if there is anything, that could be done?

**Comment by lhjnilsson** [comment]
@XuanWang-Amos any update on this?

Thanks

**Comment by sergiitk** [comment]
A quick update. We had a bit of a team reshuffle and everything got delayed, including type hints support. But the good news is that we're onboarding a new engineer at the moment, and this project is the number one on our feature support priority list.

**Comment by blankform-co** [comment]
It's a long shot but I'd love to work on this.  Is there anyone I can get in contact with? 

Would like to make sure I'm submitting high quality PRs that have a good chance of getting approved/align with your team's direction if I work on it!

**Comment by chrisprobst** [comment]
@sergiitk Any updates on this topic?

**Comment by oelhammouchi** [comment]
Hi, is there news on this?

**Comment by CLWiles-Progyny** [comment]
This might depend on the proto file, but I have been able to generate *.pyi files using --pyi-out param

Example:
```
python3 -m grpc_tools.protoc --proto_path=. pubsub_api.proto --python_out=. --pyi_out=. --grpc_python_out=.
```

**Comment by shoulder0902** [comment]
> ```
> python3 -m grpc_tools.protoc --proto_path=. pubsub_api.proto --python_out=. --pyi_out=. --grpc_python_out=.
> ```

it currently only generates *_pb2.pyi, but not *_pb2_grpc.pyi

**Comment by CLWiles-Progyny** [comment]
> > ```
> > python3 -m grpc_tools.protoc --proto_path=. pubsub_api.proto --python_out=. --pyi_out=. --grpc_python_out=.
> > ```
> 
> it currently only generates *_pb2.pyi, but not *_pb2_grpc.pyi

_grpc.py is already typed, as it contains wrapper classes to refer to grpc functions inside the untyped base proto file, which the pyi file provides type hints for. I am not seeing any type hints missing or linting errors when utilizing. _grpc.py.

**Comment by snus-kin** [comment]
> _grpc.py is already typed, as it contains wrapper classes to refer to grpc functions inside the untyped base proto file, which the pyi file provides type hints for. I am not seeing any type hints missing or linting errors when utilizing. _grpc.py.

What do you see in that file for your `xStub` classes' ctrs? I'm having issues with typing regarding the fact there doesn't seem to have a typehint associated with the channel.

Basically this, it seems to know it's a grpc.Channel but doesn't type it

```py
class ServiceStub(object):
    def __init__(self, channel):
        """Constructor.

        Args:
            channel: A grpc.Channel.
        """
        self.Endpoint = channel.stream_stream(...)
```

**Comment by CLWiles-Progyny** [comment]
> > _grpc.py is already typed, as it contains wrapper classes to refer to grpc functions inside the untyped base proto file, which the pyi file provides type hints for. I am not seeing any type hints missing or linting errors when utilizing. _grpc.py.
> 
> What do you see in that file for your `xStub` classes' ctrs? I'm having issues with typing regarding the fact there doesn't seem to have a typehint associated with the channel.
> 
> Basically this, it seems to know it's a grpc.Channel but doesn't type it
> 
> class ServiceStub(object):
>     def __init__(self, channel):
>         """Constructor.
> 
>         Args:
>             channel: A grpc.Channel.
>         """
>         self.Endpoint = channel.stream_stream(...)

I see it typed as a generic object.  As it is a function input it may be desirable for assorted reasons.

**Comment by aidandj** [comment]
Take a look at what mypy-protobuf generates for grpc: https://github.com/nipunn1313/mypy-protobuf/blob/main/test/generated/testproto/grpc/dummy_pb2_grpc.pyi

The pb2.pyi file is not sufficient for typing the arguments, or return values of stubs.

See these examples: https://github.com/aidandj/grpc-typing/blob/main/main.py

<img width="596" height="108" alt="Image" src="https://github.com/user-attachments/assets/086eae65-6585-4fed-8492-6f48bbf682dd" />

<img width="797" height="284" alt="Image" src="https://github.com/user-attachments/assets/7fbadf1d-6ec4-490e-ab58-d8cd37751765" />

As you can see, typing of the GRPC stubs makes a huge difference. It is a bummer that mypy-protobuf is no longer really being maintained,

It works for both sync and async [using overloads](https://github.com/nipunn1313/mypy-protobuf/blob/afdf55f7fbbe334ac1ed2791ef4ece3ad6724dab/test/generated/testproto/grpc/dummy_pb2_grpc.pyi#L96)

I'd probably be willing to take a stab at porting the mypy-protobuf generator to C++ if there was interest, or possibility of inclusion.

**Comment by CLWiles-Progyny** [comment]
> Take a look at what mypy-protobuf generates for grpc: https://github.com/nipunn1313/mypy-protobuf/blob/main/test/generated/testproto/grpc/dummy_pb2_grpc.pyi  ....snip....

Thanks for further clarifying, I was misreading prior comments.

Unless I am incorrect, this requires implementing another generator inside the python compiler. I'd be willing to submit a PR for this if existing contributors don't get around to it. if my C++ skills were any good, I would do it right away but maybe this is an opportunity to learn :)
